src/meta_schedule/feature_extractor/per_block_feature.cc

```diff
@@ -388,191 +383,148 @@ struct Feature {
-            auto it = var_substitutes.find(var);
-            if (it != var_substitutes.end()) {
-              return it->second;
-            }
-          }
-          return NullOpt;
-        }));
-  };
-  // Step 2.1. Substitute block vars of the parent scope to its min
-  if (!scopes->empty()) {
-    const BlockNode* parent_block = scopes->back()->block.operator->();
-    for (const IterVar& block_var : parent_block->iter_vars) {
-      var_substitutes[block_var->var.get()] = block_var->dom->min;
-    }
-  }
-  // Step 2.2. Substitute block vars to its binding
   {
+    // - 1) block vars of the parent scope are substituted with its minimal possible value
+    if (parent_scope != nullptr) {
+      for (const IterVar& block_var : parent_scope->block->iter_vars) {
+        var_substitutes[block_var->var.get()] = block_var->dom->min;
+      }
+    }
+    // - 2) block vars of the current scope are substituted with the expr bound to it
     ICHECK_EQ(realize->iter_values.size(), realize->block->iter_vars.size());
     int n = realize->iter_values.size();
     for (int i = 0; i < n; ++i) {
       const Var& lhs = realize->block->iter_vars[i]->var;
       const PrimExpr& rhs = realize->iter_values[i];
-      var_substitutes[lhs.get()] = f_substitute(rhs);
+      var_substitutes[lhs.get()] = Substitute(rhs, var_substitutes);
     }
   }
-  // Step 2.3. Helper to convert a TIR region into our int-set and do necessary simplification
-  auto f_make_int_set = [&f_substitute](const Array<Range>& region) -> NDIntSet {
+  // Step 2. A helper that turns a buffer access region into an IntSet
+  // with block var substituted and simplified.
+  auto f_make_int_set = [analyzer, &var_substitutes](const Array<Range>& region) -> NDIntSet {
     // Helper function to do the substitution
     int ndim = region.size();
     NDIntSet result;
     result.reserve(ndim);
     for (int i = 0; i < ndim; ++i) {
       const Range& range = region[i];
-      PrimExpr min = f_substitute(range->min);
-      PrimExpr max = f_substitute(min + range->extent - Integer(1));
+      PrimExpr min = analyzer->Simplify(Substitute(range->min, var_substitutes));
+      PrimExpr max = analyzer->Simplify(Substitute(min + range->extent - 1, var_substitutes));
       result.push_back(arith::IntSet::Interval(min, max));
     }
     return result;
   };
-  // Step 3. Apply the substitution to each tensor region
-  struct BufferInfo {
-    AccessType access_type = AccessType::kUnknownRW;
-    std::vector<NDIntSet> regions = {};
-  };
-  std::unordered_map<const BufferNode*, BufferInfo> buffer_info;
-  buffer_info.reserve(realize->block->reads.size() + realize->block->writes.size());
-  for (int i = 0; i < n_reads; ++i) {
-    // Skip those update regions
-    if (is_read_update[i]) {
-      continue;
+  // Step 3. For each buffer being read/written, extract the region and the access type
+  std::unordered_map<const BufferNode*, int> buffer2idx;
```

Comment:
Why do we need a map here? Looks like we only need to find if the buffer is inside the map?

Reply:
This map is used to deduplicate the buffers, as well as to find the index of a certain buffer in sub_features.

For example, suppose there is a block whose signature is:

```python
T.reads(C[...], A[...], B[...])
T.writes(C[...])
```

Then the sub-features we collected are:
- sub-features[0]: C
- sub-features[1]: A
- sub-features[2]: B

And `buffer2idx` is:

```python
{
  C: 0,
  A: 1,
  B: 2,
}
```